### ISSUE TITLE
piece_id vs color

### ISSUE THREAD

**Issue by kyidkim** [team-rook/chess-app#27]
user_id is a required field in the piece table, because it's used by the database to carry out the association between piece and user (a user has_many pieces, and a piece belongs_to a user). 

in sql terms, if we're querying for pieces belonging to a certain user in a certain game: select * from piece where user_id = 1 and game_id = 1.

storing color shouldn't be necessary because we can use the white? and black? logic inherited from the piece class. 

if we DO decide we need to commit white/black in the db, i think a better approach would be to use a boolean field to simplify evaluation and reduce room for error - so instead of comparing a string field ("white" or "black") we'd simply decide, for instance, that white is true and black false, then use true/false comparisons.

**Comment by jacobwgillespie** [comment]
My only feedback would be to:

*  pick one or the other (either a user_id or a color field) unless there's some justification for storing both
* have a particular reason for why you're doing the one and not the other

So, like if we're querying for the pieces belonging to a certain user, you could do like you've got and do a `select * from piece where user_id = 1 and game_id = 1` or you could do `select * from piece where is_white = TRUE and game_id = 1` or whatever once you know the user's color.

If you're storing the color on the pieces, when you render the HTML it will be trivial to know whether a piece is white or black, so it's easier to go from piece to color.  If you're storing the user ID on the pieces themselves, it will be easier to go from piece to user.

I guess then the question becomes which do you want to prioritize and why.  For instance, `select * from piece where user_id = 1 and game_id = 1` makes it easy to get pieces by game and user, but do we need to retrieve pieces by game and user?  Or do we always load all pieces?  Or does the individual piece need quick access to its user or color?

And there could be other concerns too - one that is not actually a concern / requirement for this app is, say, what if your opponent could drop out, returning the game back to needing a second player, then now if another player joins and we store player IDs on the pieces, then we need to update each piece (doable, but there are edge cases).  Though again, we likely don't need to worry about that specific concern.  :)

So, tl;dr - it's good to have an explicit reason why you're picking a particular design, and there's lots of different ways that this could work.  Which was probably already obvious.

👍

**Comment by cbohara** [comment]
I would love to have the database schema be as simple and straightforward as possible.  I'm feeling pretty overwhelmed.

**Comment by cbohara** [comment]
I would love to have the database schema be as simple as possible.  I'm feeling pretty overwhelmed and lost.

I don’t understand how the black_user_id and the white_user_id in the game table work.  If someone can explain this to me I would be very grateful :)

I felt like I had an aha! moment earlier today when Jacob suggested that we add a color attribute to each piece, and I would then be able to create the pieces like this example:
Rook.create(color: black, x_position: 0, y_position: 0, game_id: id)

Now I am feeling lost again.  Could we keep the color attribute for each piece?  I am totally open to making the color a boolean in which white is true and black false, then use true/false comparisons.  That is a great idea Kim.

How shall we move forward from here gang? Team ROoks!

**Comment by cbohara** [comment]
I would love to have the database schema be as simple as possible. I'm feeling pretty overwhelmed and lost.

I don’t understand how the black_user_id and the white_user_id in the game table work. If someone can explain this to me I would be very grateful :)

I felt like I had an aha! moment earlier today when Jacob suggested that we add a color attribute to each piece, and I would then be able to create the pieces like this example:
Rook.create(color: black, x_position: 0, y_position: 0, game_id: id)

Now I am feeling lost again. Could we keep the color attribute for each piece? I am totally open to making the color a boolean in which white is true and black false, then use true/false comparisons. That is a great idea Kim.

How shall we move forward from here gang? Team ROoks!

**Comment by theberbie** [comment]
I see how the user_id would be useful, but at the same time, I see more benefits in using a boolean for colors. I vote for a piece.game.color option

**Comment by jacobwgillespie** [comment]
@cbohara:

On the `black_user_id` and `white_user_id`:

So, you know how you're used to creating relationships (associations) between models?  Like, in an example app, you'd have Post and User for a blog, so you'd have:

```ruby
class User < ActiveRecord::Base
  has_many :posts
end

class Post < ActiveRecord::Base
  belongs_to :user
end
```

So, we're assuming that there are blog posts (`Post`) and users write them as their authors (`belongs_to :user`).

So, the `belongs_to :user` tells ActiveRecord that it should look for a field in the database called `user_id`, and that it should take that field to mean an ID of a User model.  It also defines a bunch of convenience methods like `def user` to fetch the user from the database `def user=(val)` to set the user, etc.

ActiveRecord "automagically" infers that if you say `belongs_to :user` it should look for a class called `User` and with the `has_many :posts` automagically infers that it should look for a field called `user_id` in the `Post` model when fetching posts, but it doesn't have to be that way.  You could, for instance, call the relationship `author`:

```ruby
class User < ActiveRecord::Base
  has_many :posts, inverse_of: :author
end

class Post < ActiveRecord::Base
  belongs_to :author, class_name: 'User'
end
```

So, now we have named the relationship author, which is kinda more semantically correct.  But the automagic Rails inferences no longer work - Posts when looking for `post.author` need to be told that the model class name is `User`, and Users need to be told that they are referred to as `author` inside the Post model.  The `belongs_to :author` will now expect that there is a database field called `author_id` in the posts table.  And we can call `post.author` instead of `post.user`.

Finally we can add multiple relationships between the same two model types.  Like suppose we decide to add an editor relationship in addition to the author one:

```ruby
class User < ActiveRecord::Base
  has_many :authored_posts, inverse_of: :author, class_name: 'User'
  has_many :edited_posts, inverse_of: :editor, class_name: 'User'
end

class Post < ActiveRecord::Base
  belongs_to :author, class_name: 'User'
  belongs_to :editor, class_name: 'User'
end
```

So, it's almost the same as duplicating the associations from before, but we had to change the name `posts` into `authored_posts` and `edited_posts` so since AuthoredPost and EditedPost are not models (those would be Rails's guess for model names), we have to explicitly tell it that the `class_name` is User.

The posts table now expects a database column called `author_id` and one called `editor_id`.  If we needed to change the column name on any of these, we could use:

```ruby
belongs_to :author, class_name: 'User', foreign_key: :user_author_id
```

So now it's looking for a database column explicitly named `user_author_id`.  No particular reason why that's necessary here, but it's possible.

All the different association docs are [here](http://guides.rubyonrails.org/association_basics.html), though there's a lot of info there so it's good as a reference.

So, in our case, we're just associating the Game and User models twice, once for `white_user` and once for `black_user`.

Hope that helps.

**Comment by cbohara** [comment]
@jacobwgillespie : I really appreciate that you took the time to write that all out.  I'm going to take some time tomorrow to wrap my head around this :)

**Comment by kyidkim** [comment]
To quote Jacob:

> If you're storing the color on the pieces, when you render the HTML it will be trivial to know whether a piece is white or black, so it's easier to go from piece to color. If you're storing the user ID on the pieces themselves, it will be easier to go from piece to user.

How are we determining who has the next move? At this point, our primary indicator of who moved most recently is the updated_at timestamp in the pieces. If we differentiate pieces by color, we still have to link it back to user - so we've just added an extra step of logic.

We already have the logic to get piece to color. However, it's not clear to me how helpful that will be when coding the front end.

I think what's really difficult about this decision is that we don't know how it will play out in future functionality.

**Comment by jacobwgillespie** [comment]
> How are we determining who has the next move?

You are correct that we don't have a way to do that yet - I'd recommend storing that on the game model though (`last_updated_by` or `next_move` or whatever), as you'd end up with a lot of database work or ruby code if you use the piece timestamp to infer that.  Plus it kinda "belongs" to the game, so makes sense to store it there.

Whichever way we choose to model things, we will need to be able to go from piece to color (at least to determine what color to show the piece on the frontend) and from piece to user (to check if a user can move that piece).  As long as we can do both of those, we're good.
